test/unit/space_index_test.lua

```diff
@@ -0,0 +1,301 @@
+        helpers.retrying({}, function()
+            t.assert_equals(helpers.iteration_result, {{1, "1"}})
+        end)
+        helpers.iteration_result = {}
+
+        local capture = luatest_capture:new()
+        capture:wrap(true, function()
+            local space = box.space[space_name]
+            case.fun(space, space.index[index_name])
+            space:insert({1, "1"})
+
+            local stderr = ""
+            helpers.retrying({}, function()
+                stderr = stderr .. capture:flush().stderr
+                t.assert_str_contains(stderr, "expiration: stop task")
+            end)
+        end)
+
+        t.assert_equals(helpers.iteration_result, {})
+        t.assert_equals(#expirationd.tasks(), 1)
+        t.assert_not_equals(expirationd.task(task_name), nil)
+        t.assert_equals(task:statistics().restarts, 1)
+
+        task:kill()
+    end
+    g["test_not_stop_after_" .. name .. "if_id"] = function(cg)
+        local task_name = "clean_all"
+        local space_id = cg.space.id
+        local index_id = 1
+
+        helpers.iteration_result = {}
+        local task = expirationd.start(task_name, space_id, helpers.is_expired_debug,
+                                       {index = index_id})
+        helpers.retrying({}, function()
+            t.assert_equals(helpers.iteration_result, {{1, "1"}})
+        end)
+        helpers.iteration_result = {}
+
+        local space = box.space[space_id]
+        case.fun(space, space.index[index_id])
+        space:insert({1, "1"})
+
+        helpers.retrying({}, function()
+            t.assert_equals(helpers.iteration_result, {{1, "1"}})
+        end)
+        t.assert_equals(#expirationd.tasks(), 1)
+        t.assert_not_equals(expirationd.task(task_name), nil)
+        t.assert_equals(task:statistics().restarts, 1)
+
+        task:kill()
+    end
+end
+
+local drop_cases = {
+    index_drop = {
+        fun = function(_, index)
+            index:drop()
+        end,
+    },
+    space_drop = {
+        fun = function(space, _)
+            space:drop()
+        end,
+    }
+}
+
+for name, case in pairs(drop_cases) do
+    g["test_stop_after_" .. name] = function(cg)
+        local task_name = "clean_all"
+        local index_id = 1
+
+        helpers.iteration_result = {}
+        local task = expirationd.start(task_name, cg.space.id, helpers.is_expired_debug,
+                                       {index = index_id})
+        helpers.retrying({}, function()
+            t.assert_equals(helpers.iteration_result, {{1, "1"}})
+        end)
+        helpers.iteration_result = {}
+
+        local capture = luatest_capture:new()
+        capture:wrap(true, function()
+            case.fun(cg.space, cg.space.index[index_id])
+
+            local stderr = ""
+            helpers.retrying({}, function()
+                stderr = stderr .. capture:flush().stderr
+                t.assert_str_contains(stderr, "expiration: stop task")
```

Comment:
So what will happen after I recreate the space or the index?

Reply:
It depends on how fast you recreate space:
1. It may stop if you were slow.
2. It may continue to work.

I've added such tests.

Idea was not to identify all wrong cases and stop, but just to stop if we can't rerun a task. If we have a space and index specified in a task options - we can rerun.

Reply:
Thank you, I'll get it.

> 2. It may continue to work if you were quick.

To be honest, I'm not sure that it should be covered with tests.

Reply:
I think it is ok, this is a test for the expected behavior.